libraries/core_libs/consensus/src/final_chain/final_chain.cpp

```diff
@@ -76,45 +77,72 @@ class FinalChainImpl final : public FinalChain {
 
   void stop() override { executor_thread_.stop(); }
 
-  future<shared_ptr<FinalizationResult const>> finalize(NewBlock new_blk, uint64_t period,
+  future<shared_ptr<FinalizationResult const>> finalize(SyncBlock&& sync_block,
                                                         finalize_precommit_ext precommit_ext = {}) override {
     auto p = make_shared<promise<shared_ptr<FinalizationResult const>>>();
-    executor_thread_.post([this, new_blk = move(new_blk), period, precommit_ext = move(precommit_ext), p]() mutable {
-      p->set_value(finalize_(move(new_blk), period, precommit_ext));
+    executor_thread_.post([this, sync_block = std::move(sync_block), precommit_ext = move(precommit_ext), p]() mutable {
+      p->set_value(finalize_(std::move(sync_block), precommit_ext));
     });
+
     return p->get_future();
   }
 
-  shared_ptr<FinalizationResult> finalize_(NewBlock new_blk, uint64_t period,
-                                           finalize_precommit_ext const& precommit_ext) {
+  shared_ptr<FinalizationResult> finalize_(SyncBlock&& sync_block, finalize_precommit_ext const& precommit_ext) {
     auto batch = db_->createWriteBatch();
-    Transactions to_execute;
-    {
-      // This artificial scope will make sure we clean up the big chunk of memory allocated for this batch-processing
-      // stuff as soon as possible
-      auto period_raw = db_->getPeriodDataRaw(period);
-      SyncBlock sync_block(period_raw);
-      DB::MultiGetQuery db_query(db_, sync_block.transactions.size());
-      for (auto const& trx : sync_block.transactions) {
-        db_query.append(DB::Columns::final_chain_transaction_location_by_hash, trx.getHash());
+
+    // Create dag & transactions stats for rewards distribution
+    RewardsStats rewards_stats(sync_block.getTransactions().size(), sync_block.getDagBlocks().size());
+    for (const auto& dag_block : sync_block.getDagBlocks()) {
+      // TODO: this assert throws in test FinalChainTest.contract, either fix the test or delete the assert
+      // assert(dag_block.getTrxs().size());
```

Comment:
I think should keep the assert here. DAG block with no transactions should be an invalid block, and need to drop

Reply:
When I think about it now, we need such condition/or assert inside proposing/receiving dag blocs and not here... This is just finalize() function that is processing rewards so at this moment such invalid dag bock is in database, dag, everywhere...